File: cses/06_tree_algorithms/1136_counting-paths/solution.py
import sys
from collections import deque


def main():
    data = sys.stdin.buffer.read().split()
    it = iter(data)
    n = int(next(it))
    m = int(next(it))

    # Đọc cây: danh sách kề vô hướng gồm n - 1 cạnh
    adj = [[] for _ in range(n + 1)]
    for _ in range(n - 1):
        a = int(next(it)); b = int(next(it))
        adj[a].append(b)
        adj[b].append(a)

    # BFS gốc tại 1: dựng parent, children và thứ tự duyệt (order)
    parent = [0] * (n + 1)
    children = [[] for _ in range(n + 1)]
    order = []
    visited = bytearray(n + 1)
    dq = deque([1])
    visited[1] = 1
    while dq:
        u = dq.popleft()
        order.append(u)
        for v in adj[u]:
            if not visited[v]:
                visited[v] = 1
                parent[v] = u
                children[u].append(v)
                dq.append(v)

    # Đọc các đường đi (truy vấn); qlist[u] = danh sách (đầu kia, chỉ số truy vấn)
    pathA = [0] * m
    pathB = [0] * m
    qlist = [[] for _ in range(n + 1)]
    for qi in range(m):
        a = int(next(it)); b = int(next(it))
        pathA[qi] = a
        pathB[qi] = b
        qlist[a].append((b, qi))
        qlist[b].append((a, qi))

    # ---- Tarjan offline LCA (DSU union theo size + nén đường đi) ----
    dsu = list(range(n + 1))
    size = [1] * (n + 1)
    ancestor = [0] * (n + 1)
    black = bytearray(n + 1)
    lca = [0] * m

    def find(x):
        root = x
        d = dsu
        while d[root] != root:
            root = d[root]
        while d[x] != root:
            nxt = d[x]
            d[x] = root
            x = nxt
        return root

    # DFS lặp: mỗi frame là [node, parent, chỉ số con kế tiếp]
    stack = [[1, 0, 0]]
    ancestor[1] = 1  # make_set(1)
    while stack:
        frame = stack[-1]
        u = frame[0]
        ci = frame[2]
        ch = children[u]
        if ci < len(ch):
            # Còn con chưa duyệt: make_set(v) rồi đi xuống con đó
            v = ch[ci]
            frame[2] = ci + 1
            ancestor[v] = v  # make_set(v)
            stack.append([v, u, 0])
        else:
            # Hoàn tất u: trả lời mọi truy vấn có đầu kia đã hoàn tất (black)
            black[u] = 1
            for w, qi in qlist[u]:
                if black[w]:
                    lca[qi] = ancestor[find(w)]
            stack.pop()
            pu = frame[1]
            if pu:
                # union(pu, u) rồi gán lại ancestor của tập cho cha pu
                ra = find(pu)
                rb = find(u)
                if ra != rb:
                    if size[ra] < size[rb]:
                        ra, rb = rb, ra
                    dsu[rb] = ra
                    size[ra] += size[rb]
                ancestor[find(pu)] = pu

    # ---- Difference on tree: đánh dấu hiệu cho từng đường đi qua LCA ----
    diff = [0] * (n + 1)
    for qi in range(m):
        a = pathA[qi]
        b = pathB[qi]
        l = lca[qi]
        diff[a] += 1
        diff[b] += 1
        diff[l] -= 1
        pl = parent[l]
        if pl:
            diff[pl] -= 1

    # ---- Cộng tổng subtree bằng cách duyệt ngược thứ tự BFS (con dồn lên cha) ----
    val = diff  # tái sử dụng mảng diff
    for i in range(len(order) - 1, -1, -1):
        u = order[i]
        p = parent[u]
        if p:
            val[p] += val[u]

    sys.stdout.buffer.write((' '.join(map(str, val[1:n + 1])) + '\n').encode())


main()

File: cses/06_tree_algorithms/1136_counting-paths/solution.cpp
#include <algorithm>
#include <iostream>
#include <utility>
#include <vector>

using namespace std;

// Cấu trúc Disjoint Set Union (union theo size + nén đường đi) dùng cho Tarjan
class DisjointSet {
public:
    explicit DisjointSet(int n) : parent_(n + 1), size_(n + 1, 1) {
        for (int node = 0; node <= n; ++node) {
            parent_[node] = node;
        }
    }

    int find(int node) {
        int root = node;
        while (parent_[root] != root) {
            root = parent_[root];
        }
        while (parent_[node] != node) {
            int next = parent_[node];
            parent_[node] = root;
            node = next;
        }
        return root;
    }

    void unite(int a, int b) {
        a = find(a);
        b = find(b);
        if (a == b) {
            return;
        }
        if (size_[a] < size_[b]) {
            swap(a, b);
        }
        parent_[b] = a;
        size_[a] += size_[b];
    }

private:
    vector<int> parent_;
    vector<int> size_;
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, path_count;
    cin >> n >> path_count;

    // Đọc cây: danh sách kề vô hướng gồm n - 1 cạnh
    vector<vector<int>> graph(n + 1);
    for (int i = 0; i < n - 1; ++i) {
        int a, b;
        cin >> a >> b;
        graph[a].push_back(b);
        graph[b].push_back(a);
    }

    // BFS lặp với gốc là đỉnh 1: dựng parent, children và thứ tự duyệt (order)
    vector<int> parent(n + 1, -1), order;
    vector<vector<int>> children(n + 1);
    order.reserve(n);
    parent[1] = 0;
    order.push_back(1);
    for (int index = 0; index < n; ++index) {
        int node = order[index];
        for (int next : graph[node]) {
            if (next == parent[node]) {
                continue;
            }
            parent[next] = node;
            children[node].push_back(next);
            order.push_back(next);
        }
    }

    // Đọc các đường đi (truy vấn LCA), lưu tại cả hai đầu để xử lý offline
    vector<int> first(path_count), second(path_count), lca(path_count);
    vector<vector<pair<int, int>>> queries(n + 1);
    for (int id = 0; id < path_count; ++id) {
        cin >> first[id] >> second[id];
        queries[first[id]].push_back({second[id], id});
        queries[second[id]].push_back({first[id], id});
    }

    // Thuật toán Tarjan offline LCA (bản lặp): DFS một lần kết hợp DSU
    DisjointSet dsu(n);
    vector<int> ancestor(n + 1), child_index(n + 1);
    vector<char> finished(n + 1, false);
    vector<int> stack = {1};
    ancestor[1] = 1;

    while (!stack.empty()) {
        int node = stack.back();
        // Còn con chưa duyệt: make_set(child) rồi đi xuống con đó
        if (child_index[node] < static_cast<int>(children[node].size())) {
            int child = children[node][child_index[node]++];
            ancestor[child] = child;
            stack.push_back(child);
            continue;
        }

        // Hoàn tất node: trả lời mọi truy vấn có đầu kia đã hoàn tất
        finished[node] = true;
        for (auto [other, id] : queries[node]) {
            if (finished[other]) {
                lca[id] = ancestor[dsu.find(other)];
            }
        }
        stack.pop_back();
        // Gộp node vào cha rồi gán lại ancestor của tập cho cha
        if (parent[node] != 0) {
            int p = parent[node];
            dsu.unite(p, node);
            ancestor[dsu.find(p)] = p;
        }
    }

    // Difference on tree: đánh dấu hiệu cho từng đường đi qua LCA của nó
    vector<long long> difference(n + 1);
    for (int id = 0; id < path_count; ++id) {
        int a = first[id];
        int b = second[id];
        int common = lca[id];
        ++difference[a];
        ++difference[b];
        --difference[common];
        if (parent[common] != 0) {
            --difference[parent[common]];
        }
    }

    // Cộng tổng subtree bằng cách duyệt ngược thứ tự BFS (con dồn lên cha)
    for (int index = n - 1; index > 0; --index) {
        int node = order[index];
        difference[parent[node]] += difference[node];
    }
    for (int node = 1; node <= n; ++node) {
        if (node > 1) {
            cout << ' ';
        }
        cout << difference[node];
    }
    cout << '\n';
}
